[#15] feat: annotate rust code blocks

jannikkeye opened the pull request:
closes #15

Please have a look at this solution, I had to split the modification and filtering of lists into to separate steps and collects since I otherwise encountered the problem described in the issue.

I will add tests when we have an acceptable and idiomatic solution.

I think I would like to somehow unify the handling of the code block state, but I am not sure how. Currently the annotation function and stripping of hidden doc tests have some code duplication.

jannikkeye commented:
I'll be looking into getting this to work with https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map to make the code cleaner hopefully today after work.

phaazon commented on src/lib.rs:
```diff
@@ -190,10 +191,15 @@ pub fn extract_inner_doc<P>(path: P, show_hidden_doc: bool, crlf: bool) -> Strin
-    .map(|line| if crlf && line == "\r\n" || line == "\n" { line } else { &line[offset..] })
+  let sanitized_annotated_lines: Vec<String> = lines
+    .into_iter()
+    .map(|line| if crlf && line == "\r\n" || line == "\n" { line } else { line[offset..].to_owned() })
```
Is this `to_owned` really mandatory? The current code benefit from not having to allocate each line on the fly with this impl:

```rust
impl<'a> FromIterator<&'a str> for String
```

You can turn your `annotate_code_blocks` function to take a `&str` as argument instead, it should work.

jannikkeye replied:
I run into lifetime issues with &str. First the compiler tells me to specify a lifetime in `annotate_code_blocks`. And I am unable to return the replaced line:

```rust
fn annotate_code_blocks<'a>(st: &mut CodeBlockState, line: &'a str) -> &'a str {
  match st {
    CodeBlockState::None => {
      if line.starts_with("~~~") {
        *st = CodeBlockState::InWithTildes;
      } else if line.starts_with("```") {
        *st = CodeBlockState::InWithBackticks;
      }

      if line == "```\n" {
        &line.replace("\n", &format!("{}\n", RUST_LANG_ANNOTATION)) //here
      } else {
        line
      }
    }

    CodeBlockState::InWithTildes => {
        if line.starts_with("~~~") {
          *st = CodeBlockState::None;
        }

        line
    }

    CodeBlockState::InWithBackticks => {
        if line.starts_with("```") {
          *st = CodeBlockState::None;
        }

        line
    }
  }
}
```

I tried it like this before and resorted to working with allocated Strings.

phaazon replied:
Whoops, sorry, I was on holidays and didn’t open GH for a while. I’ll have a closer look and think we should definitely merge that! :)

jannikkeye replied:
No problem, yea do that. Then I will resolve the other issues and add test cases. I will be on holiday starting Saturday though, hopefully I can work on this tomorrow.

phaazon commented on src/lib.rs:
````diff
@@ -289,6 +295,40 @@ fn reformat_with_markers(first_part: &str, doc: &str, second_part: &str, crlf: b
+        *st = CodeBlockState::InWithBackticks;
+      }
+
+      if line == "```\n" {
````
This will fail with Windows-style line endings.

phaazon commented on src/main.rs:
```diff
@@ -85,8 +85,14 @@
 //! This is likely to be the hidden documentation feature from `rustdoc`. To fix this problem,
 //! re-run with:
-//!
+//! 
```
Please remove this extra space.

phaazon commented on src/main.rs:
````diff
@@ -85,8 +85,14 @@
-//!
+//! 
 //! ```
+//! pub fn add(a: u8, b: u8) {
````
This function has nothing to do here.

jannikkeye commented:
@phaazon I added test cases for annotated and unannotated code blocks. Note that I split out the functionality of the `extract_inner_doc` function into two functions to allow for simpler testing.

phaazon commented:
Damn, I completely forgot about that PR. Sorry, I’m looking into it right now.

phaazon commented:
I have rebased your branch so that we can merge this and test it. :)
